File: apple/Supermessage/Previews/TextScalePreviews.swift
#if DEBUG
import SupermessageKit
import SwiftUI

/// The same screens a reader has turned their text up for.
///
/// Every frame in the catalogue was rendered at one text size, and it is the
/// default one. A reader at an accessibility size is using an application
/// nobody has looked at, and what lives there is layout: a name truncated to
/// nothing, a timestamp pushed off an edge, buttons that stop fitting side by
/// side.
///
/// Android's equivalent found exactly that on its first render — the roster
/// read `Kaa…` at twice the text size, because the name was splitting the
/// leftover space evenly with a `Spacer`. These are the frames that would
/// show the same thing here.
///
/// **`.accessibility1` and `.accessibility3`**, not the whole ladder.
/// `DynamicTypeSize` has twelve steps and they find the same bugs; one step
/// into the accessibility range and one near the top of it is where
/// two-column rows start to fight and then lose.

// The roster row: a glyph, a name, a state word, a runtime, a time and an
// unread badge, all competing for one line.
#Preview("Roster, accessibility1") {
    List {
        RoomRowView(row: PreviewFixtures.roomNeedsYou, avatarURI: nil, state: .needsYou, when: "2m")
        RoomRowView(row: PreviewFixtures.roomActive, avatarURI: nil, state: .active, when: "14m")
        RoomRowView(row: PreviewFixtures.roomInvitation, avatarURI: nil, state: .idle, when: "")
    }
    .listStyle(.plain)
    .environment(\.dynamicTypeSize, .accessibility1)
    .previewChrome()
}

#Preview("Roster, accessibility3") {
    List {
        RoomRowView(row: PreviewFixtures.roomNeedsYou, avatarURI: nil, state: .needsYou, when: "2m")
        RoomRowView(row: PreviewFixtures.roomActive, avatarURI: nil, state: .active, when: "14m")
    }
    .listStyle(.plain)
    .environment(\.dynamicTypeSize, .accessibility3)
    .previewChrome()
}

// Three buttons that have to fit a row, or stop trying to.
#Preview("Card pending, accessibility3") {
    PreviewGround {
        CustomEventCard(
            view: PreviewFixtures.cardPending, label: "Gate",
            eventType: "dev.superpipeline.gate.v1", senderName: "Superpipeline — Delivery",
            onDecide: { _ in true })
    }
    .environment(\.dynamicTypeSize, .accessibility3)
    .previewChrome()
}

// A table, a code block and a quote, none of which wrap like a paragraph.
#Preview("Rich text, accessibility3") {
    ScrollView { PreviewGround { RichTextView(blocks: PreviewFixtures.richBlocks) } }
        .environment(\.dynamicTypeSize, .accessibility3)
        .previewChrome()
}

// A message with its sender, time and reactions in one line's worth.
#Preview("Timeline row, accessibility3") {
    let media = PreviewFixtures.mediaCache()
    let faces = PreviewFixtures.faceCache()
    return VStack(alignment: .leading, spacing: 0) {
        TimelineRowView(
            row: PreviewFixtures.message, attribution: "Atlas — Platform", media: media,
            faces: faces)
        TimelineRowView(row: PreviewFixtures.withReactions, media: media, faces: faces)
    }
    .padding(.horizontal, 12)
    .frame(maxWidth: .infinity, alignment: .leading)
    .background(Theme.surface)
    .environment(\.dynamicTypeSize, .accessibility3)
    .previewChrome()
}
#endif

File: android/app/src/debug/kotlin/dev/supermessage/previews/TextScalePreviews.kt
package dev.supermessage.previews

import androidx.compose.foundation.layout.Column
import androidx.compose.runtime.Composable
import androidx.compose.ui.tooling.preview.Preview
import dev.supermessage.DecisionCard
import dev.supermessage.RichText
import dev.supermessage.RoomRow
import dev.supermessage.TimelineRow
import java.time.Instant
import uniffi.supermessage_core.AgentState

/**
 * The same screens a reader has turned their text up for.
 *
 * ## Why this file exists
 *
 * Every frame in the catalogue — 48 here, 40 on iOS, 83 web stories — was
 * rendered at **one text size**, and it is the default one. A reader at
 * `fontScale = 2.0` is using an application nobody has ever looked at, and
 * the defects that live there are layout defects: a name that truncates to
 * nothing, a row that pushes its timestamp off the edge, a card whose
 * buttons stop fitting side by side.
 *
 * Those are invisible in code review and invisible in a default-size
 * preview, which is the same combination that hid a glyph being drawn twice
 * under every agent message until someone rendered a frame.
 *
 * ## Which scales
 *
 * **1.5 and 2.0**, not the whole ladder. Android's accessibility sizes go to
 * 2.0, and a frame at each end of the range finds what a frame in the middle
 * would: 1.5 is where two-column rows start to fight, 2.0 is where they lose.
 * Every intermediate step costs a reference image and finds the same bugs.
 *
 * ## Which frames
 *
 * The ones whose layout has somewhere to break — a row with a name, a time
 * and a badge competing for one line; a card with three buttons; rich text
 * with a table in it. A frame that is one line of text at one size is one
 * line of text at every size.
 */
private val NOW: Instant = Instant.ofEpochMilli(1_757_700_120_000L)

@Preview(name = "Roster rows, 1.5x", showBackground = true, heightDp = 560, fontScale = 1.5f)
@Composable
internal fun RoomRowStatesLarge() {
    PreviewGround {
        Column {
            RoomRow(PreviewFixtures.roomNeedsYou, null, AgentState.NEEDS_YOU, "2m")
            RoomRow(PreviewFixtures.roomActive, null, AgentState.ACTIVE, "14m")
            RoomRow(PreviewFixtures.roomInvitation, null, AgentState.IDLE, "")
        }
    }
}

/**
 * The largest size Android offers, on the row that has the most to fit: a
 * glyph, a name, a state word, a runtime, a time and an unread badge.
 */
@Preview(name = "Roster rows, 2x", showBackground = true, heightDp = 720, fontScale = 2.0f)
@Composable
internal fun RoomRowStatesHuge() {
    PreviewGround {
        Column {
            RoomRow(PreviewFixtures.roomNeedsYou, null, AgentState.NEEDS_YOU, "2m")
            RoomRow(PreviewFixtures.roomActive, null, AgentState.ACTIVE, "14m")
        }
    }
}

/** Three buttons that have to fit a row, or stop trying to. */
@Preview(name = "Card, pending, 2x", showBackground = true, heightDp = 640, fontScale = 2.0f)
@Composable
internal fun CardPendingHuge() {
    PreviewGround {
        DecisionCard(
            view = PreviewFixtures.cardPending,
            label = "Gate",
            eventType = "dev.superpipeline.gate.v1",
            onDecide = { true },
        )
    }
}

/** A table, a code block and a quote, none of which wrap like a paragraph. */
@Preview(name = "Rich text, 2x", showBackground = true, heightDp = 1400, fontScale = 2.0f)
@Composable
internal fun RichTextEveryBlockHuge() {
    PreviewGround {
        RichText(blocks = PreviewFixtures.richBlocks)
    }
}

/** A message with its sender, time and reactions all on one line's worth. */
@Preview(name = "Timeline row, 2x", showBackground = true, heightDp = 520, fontScale = 2.0f)
@Composable
internal fun TimelineRowHuge() {
    PreviewGround {
        Column {
            TimelineRow(row = PreviewFixtures.message, now = NOW, attribution = "Atlas — Platform")
            TimelineRow(row = PreviewFixtures.withReactions, now = NOW, onReact = {})
        }
    }
}
